src/logger/log_collector.py

```diff
@@ -0,0 +1,232 @@
+        departures_arrivals_df = departures_arrivals_df.reset_index(drop=True)
+        return departures_arrivals_df
+
+    def _get_trains_block_section(self, run_id: UUID) -> list[str]:
+        """Returns a list of all trains that have entered or left a block
+        section in the given run.
+        :param run_id: The id of the run.
+        :return: A list of all trains that have entered or left a block"""
+
+        trains_enter = (
+            TrainEnterBlockSectionLogEntry.select(
+                TrainEnterBlockSectionLogEntry.train_id
+            )
+            .distinct()
+            .where(TrainEnterBlockSectionLogEntry.run_id == run_id)
+        )
+        trains_leave = (
+            TrainLeaveBlockSectionLogEntry.select(
+                TrainLeaveBlockSectionLogEntry.train_id
+            )
+            .distinct()
+            .where(TrainLeaveBlockSectionLogEntry.run_id == run_id)
+        )
+        train_ids = {t.train_id for t in trains_enter}
+        train_ids = train_ids.union({t.train_id for t in trains_leave})
+        return list(train_ids)
+
+    def get_block_section_times_of_train(
+        self, run_id: UUID, train_id: str
+    ) -> pd.DataFrame:
+        """Returns a DataFrame containing all block section times of the
+        given train in the given run.
+        :param run_id: The id of the run.
+        :param train_id: The id of the train.
+        :return: A DataFrame containing all block section times of the
+        given train in the given run."""
+
+        # pylint will not recognize that peewee results are iterable
+        # pylint: disable=not-an-iterable
+        train_enter_df = pd.DataFrame(
+            [
+                [e.tick, e.block_section_id, e.block_section_length]
+                for e in TrainEnterBlockSectionLogEntry.select().where(
+                    TrainEnterBlockSectionLogEntry.run_id == run_id
+                    and TrainEnterBlockSectionLogEntry.train_id == train_id
+                )
+            ],
+            columns=["tick", "block_section_id", "block_section_length"],
+        )
+        # pylint will not recognize that peewee results are iterable
+        # pylint: disable=not-an-iterable
+        train_leave_df = pd.DataFrame(
+            [
+                [e.tick, e.block_section_id, e.block_section_length]
+                for e in TrainLeaveBlockSectionLogEntry.select().where(
+                    TrainLeaveBlockSectionLogEntry.run_id == run_id
+                    and TrainLeaveBlockSectionLogEntry.train_id == train_id
+                )
+            ],
+            columns=["tick", "block_section_id", "block_section_length"],
+        )
+        train_enter_df = train_enter_df.sort_values("tick")
+        train_leave_df = train_leave_df.sort_values("tick")
+
+        train_enter_first = train_enter_df.iloc[0].tick < train_leave_df.iloc[0].tick
+        train_enter_last = train_enter_df.iloc[-1].tick >= train_leave_df.iloc[-1].tick
```

Comment:
in `get_departures_arrivals_of_train` you just used `>` compared to the `>=` here. Does this make a difference?